crates/op-rbuilder/src/integration/mod.rs

```diff
@@ -0,0 +1,182 @@
+use std::future::Future;
+use std::path::Path;
+use std::{
+    fs::{File, OpenOptions},
+    io,
+    io::prelude::*,
+    path::PathBuf,
+    process::{Child, Command},
+    time::{Duration, SystemTime},
+};
+use time::{format_description, OffsetDateTime};
+use tokio::time::sleep;
+
+/// Default JWT token for testing purposes
+pub const DEFAULT_JWT_TOKEN: &str =
+    "688f5d737bad920bdfb2fc2f488d6b6209eebda1dae949a8de91398d932c517a";
+
+mod integration_test;
+pub mod op_rbuilder;
+
+#[derive(Debug)]
+pub enum IntegrationError {
+    SpawnError,
+    BinaryNotFound,
+    SetupError,
+    LogError,
+    ServiceAlreadyRunning,
+}
+
+pub struct ServiceInstance {
+    process: Option<Child>,
+    pub log_path: PathBuf,
+}
+
+pub struct IntegrationFramework {
+    test_dir: PathBuf,
+    services: Vec<ServiceInstance>,
+}
+
+pub trait Service {
+    /// Configure and return the command to run the service
+    fn command(&self) -> Command;
+
+    /// Return a future that resolves when the service is ready
+    fn ready(&self, log_path: &Path) -> impl Future<Output = Result<(), IntegrationError>> + Send;
+}
+
+/// Helper function to poll logs periodically
+pub async fn poll_logs(
+    log_path: &Path,
+    pattern: &str,
+    interval: Duration,
+    timeout: Duration,
+) -> Result<(), IntegrationError> {
+    let start = std::time::Instant::now();
+
+    loop {
+        if start.elapsed() > timeout {
+            return Err(IntegrationError::SpawnError);
+        }
+
+        let mut file = File::open(log_path).map_err(|_| IntegrationError::LogError)?;
```

Comment:
are there concurrency issues with polling logs and services writing to the logs?

Reply:
That concurrency is handled at the os level. Worst case would be that you are reading something that has not been written yet but since it is used only for readiness checks it should be fine.